src/server/task_manager/task_manager_mixin.ts

```diff
@@ -0,0 +1,188 @@
+   * enabling the task manipulation methods, and beginning the background polling
+   * mechanism. We can't do this in a constructor due to the fact that the constructor
+   * is called before the elasticsearch plugin is available.
+   */
+  public init(kbnServer: any, server: any, config: any) {
+    this.assertUninitialized('initialize');
+
+    const logger = new TaskManagerLogger((...args: any[]) => server.log(...args));
+
+    if (!server.plugins.elasticsearch) {
+      logger.warning(
+        'The task manager cannot be initialized when the elasticsearch plugin is disabled.'
+      );
+      return;
+    }
+
+    const maxWorkers = config.get('task_manager.max_workers');
+    const definitions = extractTaskDefinitions(
+      maxWorkers,
+      kbnServer.uiExports.taskDefinitions,
+      config.get('task_manager.override_num_workers')
+    );
+    const store = new TaskStore({
+      callCluster: server.plugins.elasticsearch.getCluster('admin').callWithInternalUser,
+      index: config.get('task_manager.index'),
+      maxAttempts: config.get('task_manager.max_attempts'),
+      supportedTypes: Object.keys(definitions),
+    });
+    const pool = new TaskPool({
+      logger,
+      maxWorkers,
+    });
+    const createRunner = (instance: ConcreteTaskInstance) =>
+      new TaskManagerRunner({
+        logger,
+        kbnServer,
+        instance,
+        store,
+        definition: definitions[instance.taskType],
+        beforeRun: this.middleware.beforeRun,
+      });
+    const poller = new TaskPoller({
+      logger,
+      pollInterval: config.get('task_manager.poll_interval'),
+      work() {
+        return fillPool(pool.run, store.fetchAvailableTasks, createRunner);
+      },
+    });
+
+    server.plugins.elasticsearch.status.on('red', () => {
+      logger.debug('Lost connection to Elasticsearch, stopping the poller.');
+      poller.stop();
+    });
+
+    server.plugins.elasticsearch.status.on('green', async () => {
+      logger.debug('Initializing store');
+      await store.init();
+      logger.debug('Starting poller');
+      await poller.start();
+      logger.info('Connected to Elasticsearch, and watching for tasks');
+    });
+
+    this.store = store;
+    this.poller = poller;
+    this.isInitialized = true;
+  }
+
+  /**
+   * Adds middleware to the task manager, such as adding security layers, loggers, etc.
+   *
+   * @param {Middleware} middleware - The middlware being added.
+   */
+  public addMiddleware(middleware: Middleware) {
+    this.assertUninitialized('add middleware');
+    const prevMiddleWare = this.middleware;
+    this.middleware = addMiddlewareToChain(prevMiddleWare, middleware);
+  }
+
+  /**
+   * Schedules a task.
+   *
+   * @param task - The task being scheduled.
+   */
+  public async schedule(taskInstance: TaskInstance, options?: any) {
+    this.assertInitialized();
+    const { taskInstance: modifiedTask } = await this.middleware.beforeSave({
```

Comment:
`beforeSave` should probably be called `beforeSchedule` for better clarity

Reply:
Yes. I think that's a much better name.